test/tools/router_check/coverage.cc

```diff
@@ -69,36 +80,40 @@ void Coverage::printMissingTests(const std::set<std::string>& all_route_names,
   }
 }
 
-double Coverage::detailedReport() {
-  std::set<std::string> all_route_names;
-  std::set<std::string> covered_route_names;
-  uint64_t num_routes = 0;
-  for (const auto& host : route_config_.virtual_hosts()) {
-    for (const auto& route : host.routes()) {
-      if (route.route().has_weighted_clusters()) {
-        num_routes += route.route().weighted_clusters().clusters_size();
-      } else {
-        num_routes += 1;
-      }
-      all_route_names.emplace(route.name());
-    }
-  }
+double Coverage::getCumulativeCoverage(const std::set<std::string>& all_route_names,
+                                       const std::set<std::string>& missing_route_names) {
+  std::set<std::string> defined_covered_route_names;
+  std::set_difference(
+      all_route_names.begin(), all_route_names.end(), missing_route_names.begin(),
+      missing_route_names.end(),
+      std::inserter(defined_covered_route_names, defined_covered_route_names.end()));
   double cumulative_coverage = 0;
   for (auto& covered_route : covered_routes_) {
```

Comment:
what is the difference between `covered_routes_` and `defined_covered_route_names`?

Reply:
`covered_routes_` are all routes covered by the test files. It may include routes that do not exist in the route config, as shown in the unit tests. `defined_covered_route_names` is meant to be the intersection of routes defined by the config and routes covered by the test. I can add a comment to clarify if this approach makes sense.

Reply:
the coverage methods are called when the route matches and i think putting an extra abstraction of defined routes could indicate something is not right in the router check tool. As we discussed, can we explore the possibility of detecting direct responses in the `router.cc` and use separate checker functions for them?

Reply:
Within the existing checker functions, the logic is either based on RouteEntry or DirectResponseEntry field in the route passed in, so I modified the if-statements in those. The redirect_path coverage field already exists for DirectResponseEntry.

For future correctness tests (e.g. HTTPCode) we would need to make a change to validation.proto to add more fields for tests.

Reply:
Thank you for making the changes. These look good to me now.